server/service/client.go

```diff
@@ -294,9 +310,37 @@ func (c *Client) ApplyGroup(ctx context.Context, specs *spec.Group, logf func(fo
 	}
 
 	if len(specs.Teams) > 0 {
+		// extract the teams' custom settings and resolve the files immediately, so
+		// that any non-existing file error is found before applying the specs.
+		tmMacSettings := extractTmSpecsMacOSCustomSettings(specs.Teams)
+
+		tmFileContents := make(map[string][][]byte, len(tmMacSettings))
+		for k, paths := range tmMacSettings {
+			files := resolveMacOSCustomSettingsPaths(baseDir, paths)
+			fileContents := make([][]byte, len(files))
+			for i, f := range files {
+				b, err := os.ReadFile(f)
+				if err != nil {
+					return fmt.Errorf("applying teams: %w", err)
+				}
+				fileContents[i] = b
+			}
+			tmFileContents[k] = fileContents
+		}
+
+		// Next, apply the teams specs before saving the profiles, so that any
+		// non-existing team gets created.
 		if err := c.ApplyTeams(specs.Teams, opts); err != nil {
 			return fmt.Errorf("applying teams: %w", err)
 		}
+
+		if len(tmFileContents) > 0 {
+			for tmName, profs := range tmFileContents {
+				if err := c.ApplyTeamProfiles(tmName, profs, opts); err != nil {
+					return fmt.Errorf("applying custom settings for team %q: %w", tmName, err)
+				}
```

Comment:
(in the database I mean: since we have already applied the team specs)

Reply:
That's a possibility, those are two distinct, non-atomic operations so they may fail separately, but I wouldn't go out of my way to try and handle this edge case as there's no way to "win", i.e. the next call to clean the list of profiles may fail, and then what? There are many such non-atomic operations throughout the codebase - activities are created outside of the associated action and may fail separately, applying team specs apply each team separately, it may apply one and fail on the second, we don't "undo" the first applied team even though the `fleetctl apply` operation did fail, etc.

I'd suggest that we leave it like that, maybe to revisit later after the MDM initial release but I'd say this is a larger discussion than this specific case.

Reply:
thanks! that makes sense to me.